File: Data Structures & Algorithms/daily-temperatures/submission-0.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        output = [0] * len(temperatures)

        for i in range(len(temperatures)):
            if (not stack) or (temperatures[stack[-1]] > temperatures[i]):
                stack.append(i)
            else:
                print(stack)
                print(temperatures[i])

                while stack:
                    if temperatures[stack[-1]] < temperatures[i]:
                        output[stack[-1]] = i - stack[-1]
                        print(stack[-1])
                        stack.pop()
                    else:
                        break
                
                stack.append(i)
        
        return output

File: Data Structures & Algorithms/count-connected-components/submission-0.cpp
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> nodes(n, vector<int>{});
        vector<bool> seen(n, false);
        int count = 0;
        stack<int> st;
        
        for (auto& edge : edges) {
            nodes[edge[0]].push_back(edge[1]);
            nodes[edge[1]].push_back(edge[0]);
        }

        for (int i = 0; i < n; i++) {
            if (seen[i]) continue;

            count++;
            st.push(i);

            while (!st.empty()) {
                int node = st.top();
                st.pop();
                seen[node] = true;
                
                for (auto& child : nodes[node]) {
                    if (!seen[child]) st.push(child);
                }
            }
        }

        return count;
    }
};
